packages/material-ui/src/MenuList/MenuList.js

```diff
@@ -6,66 +6,95 @@ import ReactDOM from 'react-dom';
 import warning from 'warning';
 import ownerDocument from '../utils/ownerDocument';
 import List from '../List';
+import getScrollbarSize from 'dom-helpers/util/scrollbarSize';
+
+function MenuList({ actions, children, className, onBlur, onKeyDown, disableListWrap, ...other }) {
```

Comment:
@eps1lon and @oliviertassinari -- I'll of course adhere to the standards agreed on by the core team, but I'm curious what the formatting issues are with many props. Hoping to see the formatting ugliness that happens, I switched `List.js` to destructure props in the function declaration. After running prettier, it looked nearly identical to the previous version -- just without the `const` at the beginning or the `= props;` at the end. The indentation was completely identical. My personal preference is to destructure in the function declaration to avoid `props` being available as an alternate way to access the same variables. I also think it would be a benefit to see the defaults at the declaration of the variable rather than scrolling down to `defaultProps`. When the defaults are handled via plain JavaScript features, my IDE is also smarter about being able to instantly show me the default for the variable (a keyboard shortcut at any point of usage will show me the declaration). I'm just having a hard time seeing any downsides to destructuring in the function declaration.

Reply:
I'm personally using destructuring as few as possible as a general rule. I see two advantages. Number one, it's easier to grep. I can delete code more easily as I have more grep surface. The ability to delete code is one of my top concern when writing it. However, it's less efficient with React components. Advantage number two, in a JavaScript codebase, without TypeScript or JSDoc, you maintain the semantics of the argument, otherwise, it's implicit. However, it's less a concern in React components as it's standardized.
Regarding the `defaultProps` API. React might get ride of it https://github.com/reactjs/rfcs/pull/107. There is an issue with react-docgen for the generation of the Markdown. @eps1lon has more context than me on it.

Reply:
I'm don't have any strong preference either way. If we change the approach, my concern n°1 is consistency: do it one way everywhere, not a mix on a migration we never complete.

Reply:
> Number one, it's easier to grep.

I initially agreed with that but with modern JS `/function [A-Z]/` should be sufficient. I think this works just as well.  For arrow function components `const [A-Z].* = (` should do.

> you maintain the semantics of the argument

In general yes but for function components the argument semantics are pre-determined. There won't be a function component that starts with `context` or `ref`. Always `props`.

The reasons why I'm arguing for this is 
> My personal preference is to destructure in the function declaration to avoid props being available as an alternate way to access the same variables.

Especially in a hooks world the immediate destructuring won't even allow accessing `props.someProp` in a hook. No need for an angry linter to yell at you. Second concern why destructuring should be preferred: Minification. Minification of object properties is pretty unstable as far as I know and conservative minifaction won't touch them. However destructuring helps minification:
```diff
// with destructuring
-const { somePrettyLongName } = props;
+const { somePrettyLongName: a } = props;
-const isItReallyLong = somePrettyLongName.length > 10;
+const b = a.length > 10;
-return <div className={somePrettyLongName} />;
+return <div className={a} />;
// without
const isItReallyLong = props.somePrettyLongName.length > 10;
return <div className={props.somePrettyLongName} />;
```

I'm OK with being inconsistent with argument destructuring here since function components aren't supposed to be treated as functions anyway. Calling them directly is rather rare.

If React follows through with the deprecation of `defaultProps` we will have to adapt anyway. Enabling `react-docgen` to parse the function body for default props is probably trivial but it's time spent nonetheless.